File: src/devlane/compose.py
from __future__ import annotations

import shutil
import subprocess
from typing import Any


def build_compose_command(manifest: dict[str, Any], action: str, extra_profiles: list[str] | None = None, extra_args: list[str] | None = None) -> list[str]:
    compose_files = manifest["compose"]["files"]
    if not compose_files:
        raise ValueError("adapter does not declare any compose files")

    command = ["docker", "compose", "--env-file", manifest["paths"]["composeEnv"], "-p", manifest["network"]["projectName"]]
    for compose_file in compose_files:
        command.extend(["-f", compose_file])

    profiles = list(manifest["compose"]["profiles"])
    if extra_profiles:
        for item in extra_profiles:
            if item not in profiles:
                profiles.append(item)

    for profile in profiles:
        command.extend(["--profile", profile])

    if action == "up":
        command.extend(["up", "-d"])
    elif action == "down":
        command.append("down")
    elif action == "status":
        command.append("ps")
    else:
        raise ValueError(f"unsupported compose action: {action}")

    if extra_args:
        command.extend(extra_args)
    return command


def docker_available() -> bool:
    return shutil.which("docker") is not None


def run_compose(command: list[str], cwd: str) -> int:
    proc = subprocess.run(command, cwd=cwd, check=False)
    return int(proc.returncode)

File: internal/compose/compose.go
package compose

import (
	"fmt"
	"os"
	"os/exec"

	"github.com/auro/devlane/internal/manifest"
	"github.com/auro/devlane/internal/util"
)

func BuildCommand(manifest manifest.Manifest, action string, extraProfiles []string) ([]string, error) {
	if len(manifest.Compose.Files) == 0 {
		return nil, fmt.Errorf("adapter does not declare any compose files")
	}

	if manifest.Paths.ComposeEnv == nil {
		return nil, fmt.Errorf("compose env path is missing from manifest")
	}

	command := []string{"docker", "compose", "--env-file", *manifest.Paths.ComposeEnv, "-p", manifest.Network.ProjectName}
	for _, composeFile := range manifest.Compose.Files {
		command = append(command, "-f", composeFile)
	}

	profiles := util.DedupePreserveOrder(append(append([]string{}, manifest.Compose.Profiles...), extraProfiles...))
	for _, profile := range profiles {
		command = append(command, "--profile", profile)
	}

	switch action {
	case "up":
		command = append(command, "up", "-d")
	case "down":
		command = append(command, "down")
	case "status":
		command = append(command, "ps")
	default:
		return nil, fmt.Errorf("unsupported compose action %q", action)
	}

	return command, nil
}

func DockerAvailable() bool {
	_, err := exec.LookPath("docker")
	return err == nil
}

func Run(command []string, cwd string) int {
	cmd := exec.Command(command[0], command[1:]...)
	cmd.Dir = cwd
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin

	if err := cmd.Run(); err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return exitErr.ExitCode()
		}

		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	return 0
}
